### ISSUE TITLE
Resolution on the EIP20 API Approve / TransferFrom multiple withdrawal attack

### ISSUE THREAD

**Issue by HaleTom** [ethereum/EIPs#738]
#I raise this as a separate issue as I can't see a definitive resolution after it was [initially posted here](https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729). The thread of the issue is hard to follow because of the intermingled comments about non-related parts of the ERC20 standard.

Please close this issue if you believe that it should be posted as a comment on EIP20 itself and I will take that as a hint to move the discussion there.

But perhaps keeping separate issues for security related issues is warranted so that a single thread can be followed, months down the track. For example, the issue was raised on 29 Nov 2016, and the latest comment is on [17 June 2017](https://github.com/ethereum/EIPs/issues/20#issuecomment-309181053) -- it's nigh on impossible to follow the thread between those two dates with the interspersed other comments and references.

### Prevent EIP20 API Approve / TransferFrom multiple withdrawal attack

I seek clarity on how to effectively prevent the double withdrawal attack described in [ERC20 API: An Attack Vector on Approve/TransferFrom Methods](https://docs.google.com/document/d/1YLPtQxZu1UAvO9cZ1O2RPXBbT0mooh4DYKjA_jp-RLM/edit#)

Summary / example of the attack:
> 1. Alice allows Bob to transfer N of Alice's tokens (N>0)  by calling approve method on Token smart contract passing Bob's address and N as method arguments
>1. After some time, Alice decides to change from N to M (M>0) the number of Alice's tokens Bob is allowed to transfer, so she calls approve method again, this time passing Bob's address and M as method arguments
>1. Bob notices Alice's second transaction before it was mined and quickly sends another transaction that calls transferFrom method to transfer N Alice's tokens somewhere
>1. If Bob's transaction will be executed before Alice's transaction, then Bob will successfully transfer N Alice's tokens and will gain an ability to transfer another M tokens
>1. Before Alice noticed that something went wrong, Bob calls transferFrom method again, this time to transfer M Alice's tokens.
>
>So, Alice's attempt to change Bob's allowance from N to M (N>0 and M>0) made it possible for Bob to transfer N+M of Alice's tokens, while Alice never wanted to allow so many of her tokens to be transferred by Bob.

[EIP20 says concerning `approve()`](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20-token-standard.md#approve):
> ... clients SHOULD make sure to create user interfaces in such a way that they set the allowance first to 0 before setting it to another value for the same spender. THOUGH The contract itself shouldn't enforce it, to allow backwards compatibility with contracts deployed before

I'm confused as to why this check should be passed to the UI, and the contract shouldn't enforce it.

How is the UI to know that the `approve(spender, 0)` transaction will be processed *before* the subsequent non-zero approval, and not the other way around? It seems it is [not possible to do this check via the Web3 API](https://github.com/ethereum/EIPs/issues/20#issuecomment-263563087).

If the contract doesn't enforce this, then isn't anyone not using a UI still able to effect the attack?

I also note the lowercase `shouldn't` in "[t]he contract itself shouldn't enforce it", from which one may infer a different meaning to an [RFC2119 "SHOULD NOT"](http://www.rfc-base.org/txt/rfc-2119.txt). I've raised my issues with the wording of this paragraph [here](https://github.com/ethereum/EIPs/issues/20#issuecomment-335758020).

[EIP20 refers readers to the MiniMeToken implementation](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20-token-standard.md#implementation-of-adding-the-force-to-0-before-calling-approve-again) regarding this issue which works around it by [adding the line](https://github.com/Giveth/minime/blob/master/contracts/MiniMeToken.sol#L263):

    require((_amount == 0) || (allowed[msg.sender][_spender] == 0));

Interestingly the [nimiq-exchange-token breaks the standard with this line](https://github.com/nimiq-network/nimiq-exchange-token/commit/3f0f9599bfe66a1371a57dab59fe080f0f56ab14):

    require(_value != 0 || allowed[msg.sender][_spender] == 0);

This prevents fully cancelling an allowance after it is established (the only way would be via a `transferFrom()` of the total allowance amount)

EIP20 refers readers to two example token implementations (I've linked to the relevant line numbers):

* [OpenZeppelin](https://github.com/OpenZeppelin/zeppelin-solidity/blob/master/contracts/token/StandardToken.sol#L41) - Brief issue description and link to the original issue comment. Mentions "[o]ne possible solution to mitigate this race condition", but doesn't implement it.
* [ConsenSys](https://github.com/ConsenSys/Tokens/blob/master/contracts/StandardToken.sol#L47) - no mention or attempt to work around the issue

The [MiniMeToken function comment](https://github.com/Giveth/minime/blob/master/contracts/MiniMeToken.sol#L251) says:

> This is a modified version of the ERC20 approve function to be a little bit safer

Given this is the reference implementation, I would like to be a lot more confident than "a little bit safer".

I'm also a bit confused as to why a contract implementation is given as an exemplar when the text of EIP20 says that this should be implemented in the UI. This seems inconsistent.

Regarding [ERC223](https://github.com/ethereum/EIPs/issues/223), @[Dexaran asked](https://github.com/ethereum/EIPs/issues/20#issuecomment-289227754):
> Can someone name advantages of approve + transferFrom usage or even one thing that can not be done with fallback instead of approval.

I hope that the lack of reply (AFACT) means that fallback and the current [ERC223 recommended implementation](https://github.com/Dexaran/ERC223-token-standard/tree/Recommended) resolves this issue, but frankly I don't know enough about RFC223.

**Questions**:

1. Should the resolution to this issue be implemented in both UI and contract?
    1. What exemplar can we point to for the UI?
1. Is it possible to completely prevent this attack (and if not, why)?
1. Is the MiniMeToken approach sufficient? There was some challenge to the approach and some rebuttal, but I couldn't work out a definitive answer.
    1. If insufficient, [recommendations were made in the document describing the issue](https://docs.google.com/document/d/1YLPtQxZu1UAvO9cZ1O2RPXBbT0mooh4DYKjA_jp-RLM/edit#heading=h.6uz9seehjf3n), and also in the EIP20 comments. What is the clear direction that we should follow?
1. Could ERC223 be the point to break EIP20's attempt to ["allow backwards compatilibilty with contracts deployed before"](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20-token-standard.md#approve)? 
1. The situations in which `approve()` MAY / SHOULD / MUST return false are not currently mentioned. Does this issue necessitate the return value to be specified?

<sup>Please note that I am looking for Ethereum related work starting January 2018.</sup>

**Comment by Qqwy** [comment]
I know of two possible alternative approaches, one of which is used in the wild:

1. Add functions that increase/decrease the allowance relative to its current value, such as the [increaseApproval](https://github.com/MonolithDAO/token/blob/94b935828408d6816849797fab8f89dca9a21db3/src/Token.sol#L98) and [decreaseApproval](https://github.com/MonolithDAO/token/blob/94b935828408d6816849797fab8f89dca9a21db3/src/Token.sol#L106) functions in the MonolithDAO Token.
2. Add an alternate approval function `safeApprove(spender, new_value, expected_old_value)` that takes the current expected approval amount as input parameter. This would stop steps 3..5 in the mentioned attack from executing properly: When Bob's `transferFrom(Alice, Bob, N)` transaction occurs before Alice's `safeApprove(Bob, M, N)` transaction then the latter will fail.

**Comment by HaleTom** [comment]
It seems that the approach of adding:
``` solidity
require((_amount == 0) || (allowed[msg.sender][_spender] == 0));
```
[is not sufficient](https://github.com/OpenZeppelin/zeppelin-solidity/issues/438#issuecomment-329172399):

> 1. Bob is allowed to transfer N Alice's tokens
> 1. Alice publishes transaction that changes Bob's allowance to zero
> 1. Bob front runs Alice's transaction and transfers N Alice's tokens
> 1. Alice's transaction is mined
> 1. Alice sees that her transaction was mined successfully, that Bob's allowance is now zero and that proper Approval event was logged. This is exactly what she would see if Bob would not transfer any tokens from her, so she has no reason to think that Bob actually used his allowance before it was revoked
> 1. Now Alice publishes transaction that changes Bob's allowance to M
> 1. Alice's second transaction is mined so now Bob is allowed to transfer M Alice's tokens
> 1. Bob transfers M Alice's tokens
>
> Again. Bob got N+M tokens which is more that Alice ever wanted to allow him to transfer.
>
> One may argue that at step 5 Alice should notice that her token balance was decreased and Transfer event was logged. This is true, but if Bob had transferred tokens not to himself but to somebody else, then Transfer event will not be linked to Bob, and, if Alice's account is busy and many people are allowed to transfer from it, Alice may think that this transfer was probably performed by somebody else, not by Bob.

**Comment by HaleTom** [comment]
@Qqwy I like your safeApprove approach. All approaches are non backward compatible with ERC20 AFAICT. 

I probably should tag some previously involved people: @mikhail-vladimirov @frozeman @chriseth @Dexaran @GriffGreen @yupasik @koeppelmann @PeterBorah @jbaylina @khovratovich @brudaswen @paberr

**Comment by mikhail-vladimirov** [comment]
@HaleTom 

> All approaches are non backward compatible with ERC20 

If you just add new (safeApprove) method, this is backward compatible because all existing code will still be able to use your smart contract.

**Comment by flygoing** [comment]
@mikhail-vladimirov I think what @HaleTom meant by backwards compatible was that adding a new method means that contracts that interact with ERC20 tokens wouldn't be able to us this functionality without being updated

@HaleTom I get your issue [here](https://github.com/ethereum/EIPs/issues/738#issuecomment-336424511), and a valid solution to that is to store another bool with each allowance keeping track if any of the allowance has been used since the owner last set allowance. [Here's a quick example](https://gist.github.com/flygoing/836666010f0a5bf91abac211df938611). 
I believe this offers a few benefits over the other solutions:

1. Unless I misunderstand expectations around returning true/false from `approve`, it's backwards compatible.
2. If the spender hasn't used transferFrom since approve last succeeded, then the 1 transaction is all that's needed. The owner can tell it worked because an Approval event is broadcast with the proper amount. If it didn't work then it'll create an Approval event with 0 amount, since it set the allowance to back to 0.
3. The UI doesn't have to send a 0 approve and then the amount approve with order that matters. Now it can just send 1 if it knows the allowance hasn't been touched yet or it can just send 2 transactions, in which case the 1st to be mined will reset the `allowance` to 0/`used` to false and the 2nd will actually set the `allowance`.

Please give feedback, I'm sure I've missed something.

**Comment by mikhail-vladimirov** [comment]
The idea with boolean variable looks odd for me because it make possible the situation when `approve` method changes the allowance but returns false, while currently `false` return value means that allowance and contract's state in general was not changed.

Also this does not solve original problem.  Lets assume the following scenario:

1. Bob is allowed to transfer 100 tokens from Alice and his `used` flag is already true
2. Alice wants to change his allowance to 101
3. Alice calls `approve` method to set Bob's allowance to zero (or to 101, does not matter because Bob's `used` flag is already true)
4. Bob frontruns Alice's transaction and get 100 tokens from Alice
5. Alice sees that her transaction was mined and she has no clues to know that Bob did frontrun it
6. Now Alice calls `approve` method again to set bob's allowance to 101
7. After Alice's transaction is mined, Bob gets another 101 tokens from Alice

So Bob got 201 tokens from Alice in total while Alice never intended to give him more than 101 token.

**Comment by flygoing** [comment]
Yeah, as I pointed out, my one concern was the expectations around returning true/false from approve. I suppose you could just return true no matter what. Preferably, in any other programming language, I'd update the `used` state and then throw an error, but there's no non-reverting way to raise errors in solidity yet.

Can Alice not listen to the event emitted from her first tx? There's 3 possible scenarios in that case:

1. Alice called `approve` with a non-zero `_value` & the event `_value` matches hers. This means the previous allowance, if there was one, was never used
2. Alice called `approve` with a non-zero `_value` & the event `_value` **is** zero. This means the previous allowance was used.
3. Alice called `approve` with a zero `_value`. In this case the event `_value` will match whether or not the previous allowance as used and she wont know. It doesn't really matter since she was setting it to zero anyway.

**Comment by maurelian** [comment]
Came here from your email to consensys-diligence @HaleTom. 
> [ConsenSys](https://github.com/ConsenSys/Tokens) - no mention or attempt to work around the issue

It seems we're going to need to start adapting that to multiple token standards. We'll monitor, and if something resembling consensys is reached be happy to provide an implementation alongside our EIP20 implementation.

**Comment by HaleTom** [comment]
Has anyone written a truffle (or other) test which illustrates this issue?

**Comment by outofgas** [comment]
Hello,
if higher gas costs were acceptable one could keep track of the so-far spent allowance. Sketch:
```
    struct Allowance {
        uint initial;
        uint residual;
    }
    
    mapping(address => mapping(address => Allowance)) public allowances;
    
    function approve(address spender, uint amount) public returns (bool) {
        Allowance storage _allowance = allowances[msg.sender][spender];
        
        // This test should not be necessary.
        uint spent = _allowance.initial > _allowance.residual
                   ? _allowance.initial - _allowance.residual
                   : 0;
                   
        _allowance.initial = amount;
        _allowance.residual = spent < amount ? amount - spent : 0;
        
        Approval(msg.sender, spender, _allowance.residual);
        
        return true;
    }
    
    function allowance(address holder, address spender) public view returns (uint) {
        return allowances[holder][spender].residual;
    }
    
    function transferFrom(address holder, uint amount) public returns (bool) {
        uint residual = allowance(holder, msg.sender);
        
        require(amount <= residual);
        
        allowances[holder][msg.sender].residual = residual - amount;
        
        // ... do the token transfer
        
        return true;
    }
```
This approach would still require the allowance to be set to zero prior to setting it to a new value.

**Comment by xhyumiracle** [comment]
I think the fix mentioned [here](https://github.com/CORIONplatform/solidity/issues/107) might be helpful. Just FYI.
Personally I like the idea and it might be a great way to fix the problem, cuz I think the user who updates the allowance value should be responsible for it, (s)he should provide the latest state (tx state or allowance value) when they decide to update, otherwise the best we can do is to notice them. 
What the implementation should do is to provide a handful way for users to announce the state.

**Comment by anhlam2234** [comment]
Hey
